File: project/app/src/main/java/org/owntracks/android/model/EncryptionProvider.java
package org.owntracks.android.model;

import static org.libsodium.jni.SodiumConstants.XSALSA20_POLY1305_SECRETBOX_KEYBYTES;
import static org.libsodium.jni.SodiumConstants.XSALSA20_POLY1305_SECRETBOX_NONCEBYTES;

import android.util.Base64;

import androidx.annotation.NonNull;

import org.libsodium.jni.crypto.Random;
import org.libsodium.jni.crypto.SecretBox;
import org.owntracks.android.preferences.Preferences;

import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

import timber.log.Timber;

@Singleton
public class EncryptionProvider {
    private static final int crypto_secretbox_NONCEBYTES = XSALSA20_POLY1305_SECRETBOX_NONCEBYTES;
    private static final int crypto_secretbox_KEYBYTES = XSALSA20_POLY1305_SECRETBOX_KEYBYTES;

    private static SecretBox b;
    private static Random r;
    private static boolean enabled;
    private final Preferences preferences;

    public boolean isPayloadEncryptionEnabled() {
        return enabled;
    }

    private void initializeSecretBox() {
        String encryptionKey = preferences.getEncryptionKey();

        enabled = encryptionKey != null && !encryptionKey.isEmpty();
        Timber.v("encryption enabled: %s", enabled);
        if (!enabled) {
            return;
        }

        byte[] encryptionKeyBytes = encryptionKey != null ? encryptionKey.getBytes() : new byte[0];
        byte[] encryptionKeyBytesPadded = new byte[crypto_secretbox_KEYBYTES];

        if (encryptionKeyBytes.length == 0) {
            Timber.e("encryption key is too short or too long. Has %s bytes", encryptionKeyBytes.length);
            enabled = false;
            return;
        }
        int copyBytes = encryptionKeyBytes.length;
        if (copyBytes > crypto_secretbox_KEYBYTES) {
            copyBytes = crypto_secretbox_KEYBYTES;
        }

        System.arraycopy(encryptionKeyBytes, 0, encryptionKeyBytesPadded, 0, copyBytes);

        b = new SecretBox(encryptionKeyBytesPadded);
        r = new Random();
    }

    @Inject
    public EncryptionProvider(Preferences preferences) {
        this.preferences = preferences;
        preferences.registerOnPreferenceChangedListener(new SecretBoxManager());
        initializeSecretBox();
    }

    String decrypt(String cyphertextb64) throws Parser.EncryptionException {
        byte[] onTheWire = Base64.decode(cyphertextb64.getBytes(), Base64.DEFAULT);
        byte[] nonce = new byte[crypto_secretbox_NONCEBYTES];
        if (onTheWire.length <= crypto_secretbox_NONCEBYTES) {
            throw new Parser.EncryptionException("Message length shorter than nonce");
        }
        byte[] cyphertext = new byte[onTheWire.length - crypto_secretbox_NONCEBYTES];

        System.arraycopy(onTheWire, 0, nonce, 0, crypto_secretbox_NONCEBYTES);
        System.arraycopy(onTheWire, crypto_secretbox_NONCEBYTES, cyphertext, 0, onTheWire.length - crypto_secretbox_NONCEBYTES);
        return new String(b.decrypt(nonce, cyphertext));
    }

    String encrypt(@NonNull String plaintext) {
        return encrypt(plaintext.getBytes());
    }

    String encrypt(@NonNull byte[] plaintext) {
        byte[] nonce = r.randomBytes(crypto_secretbox_NONCEBYTES);
        byte[] cyphertext = b.encrypt(nonce, plaintext);
        byte[] out = new byte[crypto_secretbox_NONCEBYTES + cyphertext.length];

        System.arraycopy(nonce, 0, out, 0, crypto_secretbox_NONCEBYTES);
        System.arraycopy(cyphertext, 0, out, crypto_secretbox_NONCEBYTES, cyphertext.length);

        return Base64.encodeToString(out, Base64.NO_WRAP);
    }

    private class SecretBoxManager implements Preferences.OnPreferenceChangeListener {
        SecretBoxManager() {
            preferences.registerOnPreferenceChangedListener(this);
        }

        @Override
        public void onPreferenceChanged(@NonNull Set<String> properties) {
            if (properties.contains("encryptionKey")) {
                initializeSecretBox();
            }
        }
    }
}

File: project/app/src/main/java/org/owntracks/android/model/EncryptionProvider.kt
package org.owntracks.android.model

import java.security.MessageDigest
import java.security.SecureRandom
import javax.inject.Inject
import javax.inject.Singleton
import org.bouncycastle.crypto.engines.XSalsa20Engine
import org.bouncycastle.crypto.macs.Poly1305
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.params.ParametersWithIV
import org.bouncycastle.util.Arrays
import org.bouncycastle.util.encoders.Base64
import org.owntracks.android.preferences.Preferences

@Singleton
class EncryptionProvider @Inject constructor(private val preferences: Preferences) {

  var isPayloadEncryptionEnabled: Boolean = false
    private set

  private val key = ByteArray(KEYBYTES)

  init {
    preferences.registerOnPreferenceChangedListener(SecretBoxManager())
    initializeSecretBox()
  }

  private fun initializeSecretBox() {
    val keyString = preferences.encryptionKey
    isPayloadEncryptionEnabled = keyString.isNotEmpty()

    keyString.toByteArray().copyOf(KEYBYTES).copyInto(key)
  }

  fun decrypt(cypherTextBase64: String): String {
    val nonceMacCiphertext = Base64.decode(cypherTextBase64)

    // Separate nonce, mac and ciphertext
    val poly1305 = Poly1305()
    val nonce = Arrays.copyOfRange(nonceMacCiphertext, 0, NONCEBYTES)
    val mac = Arrays.copyOfRange(nonceMacCiphertext, NONCEBYTES, NONCEBYTES + poly1305.macSize)
    val ciphertext =
        Arrays.copyOfRange(
            nonceMacCiphertext, NONCEBYTES + poly1305.macSize, nonceMacCiphertext.size)

    val xSalsa20Engine =
        XSalsa20Engine().apply { init(false, ParametersWithIV(KeyParameter(key), nonce)) }

    // Generate mac key
    val macKey = ByteArray(KEYBYTES).also { xSalsa20Engine.processBytes(it, 0, it.size, it, 0) }

    // Calculate Mac
    val macCalculated =
        ByteArray(poly1305.macSize).also {
          poly1305.init(KeyParameter(macKey))
          poly1305.update(ciphertext, 0, ciphertext.size)
          poly1305.doFinal(it, 0)
        }

    // Decrypt on successful authentication
    var decrypted: ByteArray? = null
    if (MessageDigest.isEqual(macCalculated, mac)) {
      decrypted =
          ByteArray(ciphertext.size).also {
            xSalsa20Engine.processBytes(ciphertext, 0, ciphertext.size, it, 0)
          }
    }

    return decrypted?.let { String(it) } ?: "" // Or throw an exception for failed decryption
  }

  fun encrypt(plaintext: String): String {
    return encrypt(plaintext.toByteArray()) // Uses default Charset
  }

  fun encrypt(plaintext: ByteArray): String {
    // Generate random nonce
    val nonce = ByteArray(NONCEBYTES).also { SecureRandom().nextBytes(it) }

    val xSalsa20Engine =
        XSalsa20Engine().apply { init(true, ParametersWithIV(KeyParameter(key), nonce)) }

    // Generate mac key
    val macKey = ByteArray(KEYBYTES).also { xSalsa20Engine.processBytes(it, 0, it.size, it, 0) }

    // Encrypt plaintext
    val ciphertext =
        ByteArray(plaintext.size).also {
          xSalsa20Engine.processBytes(plaintext, 0, plaintext.size, it, 0)
        }

    // Generate mac
    val poly1305 = Poly1305()
    val mac =
        ByteArray(poly1305.macSize).also {
          poly1305.init(KeyParameter(macKey))
          poly1305.update(ciphertext, 0, plaintext.size) // ciphertext size = plaintext size
          poly1305.doFinal(it, 0)
        }

    // Concatenate, e.g. nonce|mac|ciphertext
    val outBytes = Arrays.concatenate(nonce, mac, ciphertext)

    return String(Base64.encode(outBytes))
  }

  private inner class SecretBoxManager : Preferences.OnPreferenceChangeListener {
    // Kotlin automatically registers the listener due to the `init` block of the outer class.
    // If this class were to be instantiated elsewhere and then registered,
    // you might need an explicit registration call in its own init or constructor.
    // For this specific structure, the outer class's init handles it.

    override fun onPreferenceChanged(properties: Set<String>) {
      if (properties.contains("encryptionKey")) {
        initializeSecretBox()
      }
    }
  }

  companion object {
    const val NONCEBYTES = 24
    const val KEYBYTES = 32
  }
}
